File: 26.py
def maxSubarraySumCircular(nums):
    total_sum = sum(nums)

    curr_max = curr_min = 0
    max_sum = float('-inf')
    min_sum = float('inf')

    for n in nums:
        curr_max = max(n, curr_max + n)
        max_sum = max(max_sum, curr_max)

        curr_min = min(n, curr_min + n)
        min_sum = min(min_sum, curr_min)

    # All numbers negative → return max_sum
    if max_sum < 0:
        return max_sum

    return max(max_sum, total_sum - min_sum)

# Example
print(maxSubarraySumCircular([1, -2, 3, -2]))  # Output: 3

File: 26.js
function maxSubarraySumCircular(nums) {
  let total = nums.reduce((a, b) => a + b, 0);
  let currMax = 0, currMin = 0;
  let maxSum = -Infinity, minSum = Infinity;

  for (let n of nums) {
    currMax = Math.max(n, currMax + n);
    maxSum = Math.max(maxSum, currMax);

    currMin = Math.min(n, currMin + n);
    minSum = Math.min(minSum, currMin);
  }

  if (maxSum < 0) return maxSum; // all negative

  return Math.max(maxSum, total - minSum);
}

console.log(maxSubarraySumCircular([1, -2, 3, -2])); // Output: 3
